### ISSUE TITLE
Warnings about nonsensical comparisons are disabled if macro expansion is involved

### ISSUE THREAD

**Issue by llunak** [llvm/llvm-project#12835]
|  |  |
| --- | --- |
| Bugzilla Link | [12463](https://llvm.org/bz12463) |
| Version | trunk |
| OS | Linux |
| Attachments | [clang patch](https://user-images.githubusercontent.com/2630040/143745040-c04a53a2-db69-4688-acdc-4db07663495c.gz), [clang patch](https://user-images.githubusercontent.com/2630040/143745041-f70e490a-4365-4ad5-aa3a-222a00624618.gz) |

## Extended Description 
```
$ cat a.cpp
```
```cpp
#define FOO "foo"
#define BAR "bar"
bool foo()
    {
    bool b;
    b = FOO == BAR;
    b = "foo" == "bar";
    return b;
    }
```
```
$ clang++ -Wall a.cpp -c
a.cpp:7:15: warning: result of comparison against a string literal is unspecified (use strncmp instead) [-Wstring-compare]
    b = "foo" == "bar";
        ~~~~~ ^
1 warning generated.                                                                                                         
```
Line 6 should generate a warning as well. As the warning says, the result of such a comparison is unspecified, so it is never correct, even if it comes from a macro expansion.

**Comment by YutongZhuu** [comment]
I believe this issue is fixed: https://godbolt.org/z/Kr1E8qfjo. We can close this issue.

**Comment by Endilll** [comment]
Appears to be fixed in Clang 20.1: https://godbolt.org/z/bvz3e3sbq
